controlled_resource_handler.go

```diff
@@ -29,23 +29,45 @@ type ControlledResourceHandler struct {
 	ControllerGvk   schema.GroupVersionKind
 }
 
+func (g *ControlledResourceHandler) enqueueMapped(metaObj meta_v1.Object, action string) {
+	name, namespace := g.getControllerNameAndNamespace(metaObj)
+	logger := g.loggerForObj(metaObj)
+
+	if name == "" {
+		if g.ControllerIndex != nil {
+			controllers, err := g.ControllerIndex.ControllerByObject(
+				metaObj.(runtime.Object).GetObjectKind().GroupVersionKind().GroupKind(), namespace, metaObj.GetName())
+			if err != nil {
+				logger.Error("Failed to get controllers for object", zap.Error(err))
+				return
+			}
+			for _, controller := range controllers {
+				controllerMeta := controller.(meta_v1.Object)
+				g.rebuildControllerByName(logger, controllerMeta.GetNamespace(), controllerMeta.GetName(), action)
+			}
+		}
+	} else {
+		g.rebuildControllerByName(logger, namespace, name, action)
+	}
+}
+
 func (g *ControlledResourceHandler) OnAdd(obj interface{}) {
-	name, namespace := g.getControllerNameAndNamespace(obj.(meta_v1.Object))
-	logger := g.loggerForObj(obj.(meta_v1.Object))
-	g.rebuildControllerByName(logger, namespace, name, "added")
+	metaObj := obj.(meta_v1.Object)
+	g.enqueueMapped(metaObj, "added")
 }
 
 func (g *ControlledResourceHandler) OnUpdate(oldObj, newObj interface{}) {
-	oldName, oldNamespace := g.getControllerNameAndNamespace(oldObj.(meta_v1.Object))
+	oldMeta := oldObj.(meta_v1.Object)
+	newMeta := newObj.(meta_v1.Object)
 
-	newName, newNamespace := g.getControllerNameAndNamespace(newObj.(meta_v1.Object))
+	oldName, _ := g.getControllerNameAndNamespace(oldMeta)
+	newName, _ := g.getControllerNameAndNamespace(newMeta)
 
-	if oldName != newName { // changed controller of the object
-		logger := g.loggerForObj(oldObj.(meta_v1.Object))
-		g.rebuildControllerByName(logger, oldNamespace, oldName, "updated")
+	if oldName != newName {
```

Comment:
Hm, what if controller name is the same but now the namespace is different? Should be `if oldName != newName || oldNamespace != newNamespace`, right?

Reply:
This can't be the case because "getControllerAndNamespace" always returns the same namespace of the object. The only way this can happen is if the namespace of the object changed. Owner References have to be in the same namespace as the object.